File: android/core/src/main/java/net/sw/browser/Library.kt
package net.sw.browser

import android.content.ContentUris
import android.content.ContentValues
import android.content.Context
import android.net.Uri
import android.os.Build
import android.os.Environment
import android.provider.MediaStore
import java.io.File

/**
 * What has been saved, and what can be done with it.
 *
 * Videos land in `Movies/Shard` and music in `Music/Shard`, where the gallery
 * and music apps can see them and any player can open them — that is the point
 * of putting them there rather than in the app's own storage. So the library
 * does not keep a database of its own: it asks the system what is in those
 * folders every time it is opened. A file deleted from a file manager is then
 * simply gone from here too, rather than being a row that opens nothing.
 *
 * Folders are real directories underneath it. Grouping is therefore something
 * the phone keeps, not something only this app understands — move a file out
 * with a file manager and the grouping moves with it.
 */
object Library {

    /** The app's own corner of each public folder: `Movies/Shard`, `Music/Shard`. */
    const val ROOT = "Shard"

    /**
     * The public downloads folder, by name.
     *
     * Spelled out rather than read from `Environment.DIRECTORY_DOWNLOADS`. The
     * value is the same — it is fixed by the platform and is the string the
     * media store keeps in every row — but the framework field is not a
     * compile-time constant, so off a phone it reads as null and every path
     * built from it comes out wrong in a way nothing complains about.
     */
    const val DOWNLOADS = "Download"

    /**
     * The public movies folder, by name, for the same reason as [DOWNLOADS].
     *
     * Videos are saved under `Movies/Shard` — inside the folder the gallery and
     * every video app already watch, so a download shows up beside the phone's
     * own videos, and in a corner of it that is only ours, so the library lists
     * what this app saved and not the phone's whole camera roll.
     */
    const val MOVIES = "Movies"

    /**
     * The public music folder, by name, for the same reason as [DOWNLOADS].
     *
     * Music of its own is saved under `Music/Shard` — inside the folder every
     * music app already watches, so it is found beside everything else, and in
     * a corner of it that is only ours, so the library can list what this app
     * saved without dragging in the phone's whole music collection.
     */
    const val MUSIC = "Music"

    /** Whether a saved file is a video or a music-only track. */
    enum class Kind { VIDEO, MUSIC }

    /** One saved file. */
    data class Item(
        val id: Long,
        val uri: Uri,
        val name: String,
        /** The folder it sits in, or "" for the top of the library. */
        val folder: String,
        val bytes: Long,
        /** When it was saved, in seconds since the epoch. */
        val addedAt: Long,
        /** Video or music — what shelf it belongs on. */
        val kind: Kind = Kind.VIDEO,
    ) {
        /** The name without its extension, which is what a title should read as. */
        val title: String get() = name.substringBeforeLast('.', name)
    }

    // ---- reading -----------------------------------------------------------

    /**
     * Everything saved, newest first.
     *
     * Blocking: it queries the media store. Call it off the main thread, the
     * same as anything else that touches storage.
     */
    fun items(context: Context): List<Item> =
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            (fromMediaStore(context) + musicFromMediaStore(context))
                .sortedByDescending { it.addedAt }
        } else {
            fromDisk()
        }

    /**
     * The folders in use, plus any that were made and are still empty.
     *
     * An empty folder does not exist on disk — a directory with nothing in it is
     * not something the media store will report — so the names of folders made
     * but not yet filled are remembered separately. They stop being remembered
     * once something is in them, because by then the folder speaks for itself.
     */
    fun folders(context: Context, items: List<Item>, kind: Kind): List<String> {
        val used = items.filter { it.kind == kind && it.folder.isNotBlank() }.map { it.folder }.toSet()
        return (used + remembered(context, kind)).sorted()
    }

    /** Remember a folder name so it can be shown before anything is in it. */
    fun addFolder(context: Context, name: String, kind: Kind) {
        val clean = clean(name)
        if (clean.isBlank()) return
        prefs(context).edit().putStringSet(emptyKey(kind), remembered(context, kind) + clean).apply()
    }

    /**
     * Forget a folder, moving anything inside it back to the top.
     *
     * Deleting the files with it would be the other reading of "remove this
     * folder", and it is the one that cannot be undone — so it is not the one
     * this does.
     */
    fun removeFolder(context: Context, name: String, items: List<Item>, kind: Kind) {
        items.filter { it.kind == kind && it.folder == name }.forEach { move(context, it, "") }
        prefs(context).edit().putStringSet(emptyKey(kind), remembered(context, kind) - name).apply()
    }

    // ---- changing ----------------------------------------------------------

    /**
     * Put an item in a folder, or back at the top when [folder] is blank.
     *
     * Returns false when the phone would not do it, which happens on the
     * versions where a file's location is not something an app may change.
     */
    fun move(context: Context, item: Item, folder: String): Boolean {
        val clean = clean(folder)
        if (item.folder == clean) return true
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            val values = ContentValues().apply {
                put(MediaStore.MediaColumns.RELATIVE_PATH, pathFor(item.kind, clean))
            }
            runCatching { context.contentResolver.update(item.uri, values, null, null) > 0 }
                .getOrDefault(false)
        } else {
            val root = if (item.kind == Kind.MUSIC) musicLegacyRoot() else legacyRoot()
            val target = File(root, clean).apply { mkdirs() }
            val from = File(item.uri.path ?: return false)
            runCatching { from.renameTo(File(target, item.name)) }.getOrDefault(false)
        }
    }

    /** Remove a saved file for good. */
    fun delete(context: Context, item: Item): Boolean =
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            runCatching { context.contentResolver.delete(item.uri, null, null) > 0 }
                .getOrDefault(false)
        } else {
            runCatching { File(item.uri.path ?: return false).delete() }.getOrDefault(false)
        }

    // ---- naming ------------------------------------------------------------

    /**
     * A folder name the file system will accept.
     *
     * Separators are the ones that matter: a name carrying one would put the
     * folder somewhere else entirely, which is not what typing it meant.
     */
    fun clean(name: String): String =
        name.trim().replace(Regex("""[\\/:*?"<>|]"""), "").take(40)

    /** Where a folder sits, as the media store spells it: under `Movies/Shard`. */
    fun relativePath(folder: String): String {
        val base = "$MOVIES/$ROOT"
        return if (folder.isBlank()) "$base/" else "$base/$folder/"
    }

    /** Where music sits, as the media store spells it: under `Music/Shard`. */
    fun musicPath(folder: String): String {
        val base = "$MUSIC/$ROOT"
        return if (folder.isBlank()) "$base/" else "$base/$folder/"
    }

    /** Where a folder sits for a given shelf: videos under Movies, music under Music. */
    fun pathFor(kind: Kind, folder: String): String =
        if (kind == Kind.MUSIC) musicPath(folder) else relativePath(folder)

    /**
     * The folder out of a stored path.
     *
     * The media store reports where a file is as a trailing path — the same
     * string that was written when it was put there. Everything after the
     * library's own root, and before the file, is the folder; nothing there
     * means the top of the library.
     */
    fun folderOf(relativePath: String): String {
        return folderIn(relativePath, Kind.VIDEO)
    }

    /** The folder out of a stored path, for the shelf it belongs to. */
    private fun folderIn(relativePath: String, kind: Kind): String {
        val marker = if (kind == Kind.MUSIC) "$MUSIC/$ROOT" else "$MOVIES/$ROOT"
        val at = relativePath.indexOf(marker)
        if (at < 0) return ""
        return relativePath.substring(at + marker.length).trim('/')
    }

    // ---- where it is kept --------------------------------------------------

    private const val PREFS = "shard-library"
    private const val EMPTY_FOLDERS = "empty-folders"
    private const val EMPTY_FOLDERS_MUSIC = "empty-folders-music"
    private const val BACKGROUND = "background-playback"
    // The two booleans an earlier build used, kept only so a setting already
    // made can be migrated to [END_ACTION] once.
    private const val AUTOPLAY = "autoplay-next"
    private const val SHUFFLE = "shuffle-play"
    private const val END_ACTION = "playback-end"

    private fun prefs(context: Context) =
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    /**
     * Whether sound carries on when the app is put away.
     *
     * Off by default. A video that keeps playing after the phone is put in a
     * pocket is a surprise, and a surprise that costs battery — so it is a
     * choice rather than a behaviour, and the choice is remembered.
     */
    fun backgroundPlayback(context: Context): Boolean =
        prefs(context).getBoolean(BACKGROUND, false)

    fun setBackgroundPlayback(context: Context, on: Boolean) {
        prefs(context).edit().putBoolean(BACKGROUND, on).apply()
    }

    /**
     * What happens when a video ends.
     *
     * One choice rather than two switches. "Keep playing" and "play at random"
     * are not two independent things to turn on — the first is *whether* to go
     * on, the second is *in what order*, and a checkbox each leaves the nonsense
     * state of "random order, but do not go on". A single choice of what the end
     * of a video means has exactly one answer and no nonsense state.
     */
    enum class PlaybackEnd { STOP, NEXT, SHUFFLE }

    /**
     * The chosen end action, defaulting to stop.
     *
     * Migrated once from the two booleans an earlier build stored, so a setting
     * already made carries over: shuffle wins, then next, then stop.
     */
    fun playbackEnd(context: Context): PlaybackEnd {
        val stored = prefs(context).getString(END_ACTION, null)
        if (stored != null) {
            return runCatching { PlaybackEnd.valueOf(stored) }.getOrDefault(PlaybackEnd.STOP)
        }
        val migrated = when {
            prefs(context).getBoolean(SHUFFLE, false) -> PlaybackEnd.SHUFFLE
            prefs(context).getBoolean(AUTOPLAY, false) -> PlaybackEnd.NEXT
            else -> PlaybackEnd.STOP
        }
        setPlaybackEnd(context, migrated)
        return migrated
    }

    fun setPlaybackEnd(context: Context, value: PlaybackEnd) {
        prefs(context).edit().putString(END_ACTION, value.name).apply()
    }

    private fun remembered(context: Context, kind: Kind): Set<String> =
        prefs(context).getStringSet(emptyKey(kind), emptySet()) ?: emptySet()

    private fun emptyKey(kind: Kind) =
        if (kind == Kind.MUSIC) EMPTY_FOLDERS_MUSIC else EMPTY_FOLDERS

    private fun legacyRoot() =
        File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_MOVIES), ROOT)

    private fun musicLegacyRoot() =
        File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_MUSIC), ROOT)

    /** The videos, from the video collection under `Movies/Shard`. */
    @androidx.annotation.RequiresApi(Build.VERSION_CODES.Q)
    private fun fromMediaStore(context: Context): List<Item> {
        // Everything under the library's root, at any depth. The trailing
        // wildcard is what picks up the folders inside it.
        return queryStore(
            context,
            MediaStore.Video.Media.getContentUri(MediaStore.VOLUME_EXTERNAL_PRIMARY),
            "$MOVIES/$ROOT/%",
            Kind.VIDEO,
        )
    }

    /**
     * The music, from the audio collection under `Music/Shard`.
     *
     * Only this app's own tracks come back even without a media permission: an
     * app can always read the media it created, and the query is narrowed to
     * our own corner of Music besides — so the phone's other music never shows
     * up here.
     */
    @androidx.annotation.RequiresApi(Build.VERSION_CODES.Q)
    private fun musicFromMediaStore(context: Context): List<Item> {
        return queryStore(
            context,
            MediaStore.Audio.Media.getContentUri(MediaStore.VOLUME_EXTERNAL_PRIMARY),
            "$MUSIC/$ROOT/%",
            Kind.MUSIC,
        )
    }

    @androidx.annotation.RequiresApi(Build.VERSION_CODES.Q)
    private fun queryStore(context: Context, collection: Uri, like: String, kind: Kind): List<Item> {
        val columns = arrayOf(
            MediaStore.MediaColumns._ID,
            MediaStore.MediaColumns.DISPLAY_NAME,
            MediaStore.MediaColumns.RELATIVE_PATH,
            MediaStore.MediaColumns.SIZE,
            MediaStore.MediaColumns.DATE_ADDED,
        )
        val where = "${MediaStore.MediaColumns.RELATIVE_PATH} LIKE ?"
        val args = arrayOf(like)
        val order = "${MediaStore.MediaColumns.DATE_ADDED} DESC"

        val found = mutableListOf<Item>()
        runCatching {
            context.contentResolver.query(collection, columns, where, args, order)?.use { cursor ->
                val id = cursor.getColumnIndexOrThrow(MediaStore.MediaColumns._ID)
                val name = cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.DISPLAY_NAME)
                val path = cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.RELATIVE_PATH)
                val size = cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.SIZE)
                val added = cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.DATE_ADDED)
                while (cursor.moveToNext()) {
                    val row = cursor.getLong(id)
                    found += Item(
                        id = row,
                        uri = ContentUris.withAppendedId(collection, row),
                        name = cursor.getString(name).orEmpty(),
                        folder = folderIn(cursor.getString(path).orEmpty(), kind),
                        bytes = cursor.getLong(size),
                        addedAt = cursor.getLong(added),
                        kind = kind,
                    )
                }
            }
        }
        return found
    }

    private fun fromDisk(): List<Item> {
        val found = mutableListOf<Item>()
        fun sweep(root: File, kind: Kind) {
            if (!root.isDirectory) return
            fun walk(dir: File, folder: String) {
                dir.listFiles()?.forEach { file ->
                    when {
                        file.isDirectory -> walk(file, file.name)
                        file.isFile -> found += Item(
                            id = file.absolutePath.hashCode().toLong(),
                            uri = Uri.fromFile(file),
                            name = file.name,
                            folder = folder,
                            bytes = file.length(),
                            addedAt = file.lastModified() / 1000,
                            kind = kind,
                        )
                    }
                }
            }
            walk(root, "")
        }
        sweep(legacyRoot(), Kind.VIDEO)
        sweep(musicLegacyRoot(), Kind.MUSIC)
        return found.sortedByDescending { it.addedAt }
    }
}

File: crates/shard/src/library.rs
//! What has been saved, and what can be done with it.
//!
//! The desktop half of the offline library. Downloads land in the user's own
//! `Videos\Shard` and `Music\Shard`, where Explorer and every player can already
//! see them — so this keeps no database of its own and asks the folders what is
//! in them each time it is opened. A file deleted from Explorer is then simply
//! gone from here too, rather than being a row that opens nothing.
//!
//! Folders are real directories underneath those two. Grouping is therefore
//! something the file system keeps, not something only this program understands:
//! move a file out with Explorer and the grouping moves with it.

use std::path::{Path, PathBuf};

/// The app's own corner of each of the user's media folders.
pub const ROOT: &str = "Shard";

/// Which shelf a saved file belongs on.
#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum Kind {
    Video,
    Music,
}

impl Kind {
    pub const ALL: [Kind; 2] = [Kind::Video, Kind::Music];

    pub fn label(self) -> &'static str {
        match self {
            Kind::Video => "영상",
            Kind::Music => "음악",
        }
    }

    /// Where this shelf keeps its files.
    pub fn folder(self) -> PathBuf {
        let base = std::env::var("USERPROFILE")
            .map(PathBuf::from)
            .unwrap_or_else(|_| std::env::temp_dir());
        match self {
            Kind::Video => base.join("Videos").join(ROOT),
            Kind::Music => base.join("Music").join(ROOT),
        }
    }
}

/// One saved file.
#[derive(Clone, Debug)]
pub struct Item {
    pub path: PathBuf,
    /// The file's name without its extension, which is what a title reads as.
    pub title: String,
    /// The folder it sits in, or empty for the top of the shelf.
    pub folder: String,
    pub bytes: u64,
    /// When it was saved, as seconds since the epoch.
    pub saved_at: u64,
    pub kind: Kind,
}

/// Everything on a shelf, newest first.
///
/// Blocking: it reads directories. Cheap enough for a folder of downloads, and
/// it is only done when the library is opened or something changes under it.
pub fn items(kind: Kind) -> Vec<Item> {
    let root = kind.folder();
    let mut found = Vec::new();
    sweep(&root, "", kind, &mut found);
    // Newest first: what was just saved is what is being looked for.
    found.sort_by(|a, b| b.saved_at.cmp(&a.saved_at));

    // Then whatever arrangement the user has made of it. Anything they have not
    // placed — a download that landed since — keeps its place at the top rather
    // than falling to the end of a list they arranged before it existed.
    let arranged = order(kind);
    if !arranged.is_empty() {
        found.sort_by_key(|item| match arranged.iter().position(|k| *k == key(&item.path)) {
            Some(at) => (1, at),
            None => (0, 0),
        });
    }
    found
}

/// One level of folders, and the files in each — the same shape the phone keeps.
fn sweep(dir: &Path, folder: &str, kind: Kind, out: &mut Vec<Item>) {
    let Ok(entries) = std::fs::read_dir(dir) else { return };
    for entry in entries.flatten() {
        let path = entry.path();
        let Ok(meta) = entry.metadata() else { continue };
        if meta.is_dir() {
            // Only the one level: a folder inside a folder is more places to
            // look through than a list of downloads is worth.
            if folder.is_empty() {
                let name = entry.file_name().to_string_lossy().to_string();
                sweep(&path, &name, kind, out);
            }
            continue;
        }
        // Whatever is in these folders is ours to show; a stray file the user
        // dropped in plays as well as one this program saved.
        let title = path
            .file_stem()
            .map(|s| s.to_string_lossy().to_string())
            .unwrap_or_default();
        if title.is_empty() {
            continue;
        }
        out.push(Item {
            path,
            title,
            folder: folder.to_string(),
            bytes: meta.len(),
            saved_at: meta
                .modified()
                .ok()
                .and_then(|t| t.duration_since(std::time::UNIX_EPOCH).ok())
                .map(|d| d.as_secs())
                .unwrap_or(0),
            kind,
        });
    }
}

/// The folders on a shelf, in the order they should be shown.
///
/// Read from the disk rather than gathered from the files: a folder just made
/// has nothing in it yet, and one that only existed where its files said it did
/// would not appear until something was put in it — which is exactly when it is
/// least useful.
pub fn folders(kind: Kind) -> Vec<String> {
    let Ok(entries) = std::fs::read_dir(kind.folder()) else { return Vec::new() };
    let mut names: Vec<String> = entries
        .flatten()
        .filter(|e| e.metadata().map(|m| m.is_dir()).unwrap_or(false))
        .map(|e| e.file_name().to_string_lossy().to_string())
        .collect();
    names.sort();
    names
}

/// Make a folder on a shelf. Returns false when the name is unusable or the
/// file system refuses it.
pub fn add_folder(kind: Kind, name: &str) -> bool {
    let clean = clean(name);
    if clean.is_empty() {
        return false;
    }
    std::fs::create_dir_all(kind.folder().join(clean)).is_ok()
}

/// A name for a file that means the same file next time the program runs.
///
/// FNV-1a over its path: a few lines, no dependency, and far better spread than
/// a sum for what is only a lookup key.
pub fn key(path: &Path) -> String {
    let mut hash: u64 = 0xcbf2_9ce4_8422_2325;
    for byte in path.to_string_lossy().as_bytes() {
        hash ^= *byte as u64;
        hash = hash.wrapping_mul(0x1000_0000_01b3);
    }
    format!("{hash:016x}")
}

/// Where a shelf's hand-made order is written down.
///
/// Beside the config rather than among the media: the user's Videos folder is
/// theirs, and a file of ours in it is something they would have to wonder
/// about. Losing it costs the arrangement, not the files.
fn order_file(kind: Kind) -> PathBuf {
    let name = match kind {
        Kind::Video => "order-video.txt",
        Kind::Music => "order-music.txt",
    };
    uikit::config::app_dir(crate::config::APP_NAME).join(name)
}

/// The order the shelf was last arranged in, oldest arrangement first.
pub fn order(kind: Kind) -> Vec<String> {
    std::fs::read_to_string(order_file(kind))
        .map(|text| text.lines().map(|line| line.trim().to_string()).filter(|l| !l.is_empty()).collect())
        .unwrap_or_default()
}

/// Write down an arrangement, by the keys of the files in it.
pub fn set_order(kind: Kind, keys: &[String]) -> anyhow::Result<()> {
    let path = order_file(kind);
    if let Some(parent) = path.parent() {
        std::fs::create_dir_all(parent)?;
    }
    std::fs::write(path, keys.join("\n"))?;
    Ok(())
}

/// Take a folder off a shelf, keeping everything that was in it.
///
/// What is inside comes out to the top of the shelf first, and only the empty
/// folder itself is removed. Deleting a folder is a tidying-up gesture — the
/// files in it are hours of downloading, and nobody means to lose them by
/// putting a folder away. Anything already at the top with that name is left
/// alone and the file inside keeps a number on the end.
///
/// Returns how many files came out.
pub fn drop_folder(kind: Kind, name: &str) -> anyhow::Result<usize> {
    use anyhow::Context;
    let clean = clean(name);
    if clean.is_empty() {
        anyhow::bail!("이름이 비어 있습니다");
    }
    let root = kind.folder();
    let folder = root.join(&clean);
    if !folder.is_dir() {
        anyhow::bail!("{} 폴더가 없습니다", folder.display());
    }
    let mut moved = 0;
    for entry in std::fs::read_dir(&folder).with_context(|| format!("reading {}", folder.display()))? {
        let Ok(entry) = entry else { continue };
        let from = entry.path();
        if !from.is_file() {
            continue;
        }
        let Some(file) = from.file_name() else { continue };
        if std::fs::rename(&from, free_name(&root, file)).is_ok() {
            moved += 1;
        }
    }
    // Only when it is empty. A folder that still holds something — a file that
    // could not be moved, a folder of its own — is left where it is rather than
    // taken away with what is inside it.
    std::fs::remove_dir(&folder).with_context(|| format!("removing {}", folder.display()))?;
    Ok(moved)
}

/// `name` inside `root`, with a number added if something is already there.
fn free_name(root: &std::path::Path, name: &std::ffi::OsStr) -> std::path::PathBuf {
    let first = root.join(name);
    if !first.exists() {
        return first;
    }
    let name = std::path::Path::new(name);
    let stem = name.file_stem().unwrap_or(name.as_os_str()).to_string_lossy().to_string();
    let extension = name.extension().map(|e| e.to_string_lossy().to_string());
    for n in 2..1000 {
        let tried = match &extension {
            Some(extension) => root.join(format!("{stem} ({n}).{extension}")),
            None => root.join(format!("{stem} ({n})")),
        };
        if !tried.exists() {
            return tried;
        }
    }
    first
}

/// Put an item in a folder, or back at the top of its shelf when [folder] is
/// empty. The file is moved on disk, so Explorer sees the same thing this does.
pub fn move_to(item: &Item, folder: &str) -> bool {
    let clean = clean(folder);
    if item.folder == clean {
        return true;
    }
    let mut target = item.kind.folder();
    if !clean.is_empty() {
        target = target.join(&clean);
        if std::fs::create_dir_all(&target).is_err() {
            return false;
        }
    }
    let Some(name) = item.path.file_name() else { return false };
    std::fs::rename(&item.path, target.join(name)).is_ok()
}

/// Delete a saved file for good.
pub fn delete(item: &Item) -> bool {
    std::fs::remove_file(&item.path).is_ok()
}

/// Give a saved file another name, keeping the extension it already has.
///
/// The extension is not the user's to type: it says what the file is, and a
/// title typed over it would leave a video that nothing will open. Only the
/// part that reads as a title changes.
pub fn rename(item: &Item, title: &str) -> bool {
    let clean = clean(title);
    if clean.is_empty() || clean == item.title {
        return false;
    }
    let target = match item.path.extension().and_then(|e| e.to_str()) {
        Some(extension) => item.path.with_file_name(format!("{clean}.{extension}")),
        None => item.path.with_file_name(clean),
    };
    if target.exists() {
        return false;
    }
    std::fs::rename(&item.path, target).is_ok()
}

/// A folder name the file system will take.
///
/// Separators are what matter: a name carrying one would put the folder
/// somewhere else entirely, which is not what typing it meant.
pub fn clean(name: &str) -> String {
    name.trim()
        .chars()
        .filter(|c| !matches!(c, '\\' | '/' | ':' | '*' | '?' | '"' | '<' | '>' | '|'))
        .take(40)
        .collect::<String>()
        .trim()
        .to_string()
}

/// Bytes, as a person reads them.
pub fn human(bytes: u64) -> String {
    const GB: u64 = 1 << 30;
    const MB: u64 = 1 << 20;
    match bytes {
        b if b >= GB => format!("{:.1} GB", b as f64 / GB as f64),
        b if b >= MB => format!("{} MB", b / MB),
        b => format!("{} KB", (b / 1024).max(1)),
    }
}

/// How long ago it was saved, in the words a list wants.
pub fn age(saved_at: u64) -> String {
    let now = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_secs())
        .unwrap_or(0);
    if saved_at == 0 || saved_at > now {
        return String::new();
    }
    let days = (now - saved_at) / 86_400;
    match days {
        0 => "오늘".into(),
        1 => "어제".into(),
        d if d < 30 => format!("{d}일 전"),
        d => format!("{}개월 전", d / 30),
    }
}

/// Hand a file to whatever the user opens that kind with.
///
/// Their own player, not one built into this program: a media player is a large
/// thing to write badly, and the one they have chosen already knows their
/// subtitles, their volume, and where they left off.
pub fn open(path: &Path) -> bool {
    open_with_shell(path.as_os_str())
}

/// Open a folder in Explorer.
pub fn reveal(path: &Path) -> bool {
    open_with_shell(path.as_os_str())
}

#[cfg(windows)]
fn open_with_shell(what: &std::ffi::OsStr) -> bool {
    use std::os::windows::ffi::OsStrExt;

    // ShellExecuteW rather than `cmd /C start`: no console window flashes up,
    // and a path with a space or an ampersand in it needs no quoting rules
    // guessed at. Names here are Korean video titles — exactly the case a
    // command line mangles.
    let wide: Vec<u16> = what.encode_wide().chain(std::iter::once(0)).collect();
    let verb: Vec<u16> = "open\0".encode_utf16().collect();
    #[link(name = "shell32")]
    unsafe extern "system" {
        fn ShellExecuteW(
            hwnd: isize,
            operation: *const u16,
            file: *const u16,
            parameters: *const u16,
            directory: *const u16,
            show: i32,
        ) -> isize;
    }
    // Anything above 32 is success; below is one of the shell's error codes.
    const SW_SHOWNORMAL: i32 = 1;
    unsafe {
        ShellExecuteW(
            0,
            verb.as_ptr(),
            wide.as_ptr(),
            std::ptr::null(),
            std::ptr::null(),
            SW_SHOWNORMAL,
        ) > 32
    }
}

#[cfg(not(windows))]
fn open_with_shell(_what: &std::ffi::OsStr) -> bool {
    false
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_folder_name_loses_what_a_path_would_take_as_a_separator() {
        assert_eq!(clean("여행/2024"), "여행2024");
        assert_eq!(clean("  운동  "), "운동");
        assert_eq!(clean("a:b*c?"), "abc");
    }

    #[test]
    fn a_folder_name_is_kept_to_a_length_a_list_can_show() {
        assert_eq!(clean(&"가".repeat(80)).chars().count(), 40);
    }

    #[test]
    fn the_two_shelves_are_the_users_own_media_folders() {
        let video = Kind::Video.folder();
        let music = Kind::Music.folder();
        assert!(video.ends_with(std::path::Path::new("Videos").join(ROOT)));
        assert!(music.ends_with(std::path::Path::new("Music").join(ROOT)));
    }

    #[test]
    fn a_file_coming_out_of_a_folder_does_not_land_on_one_already_there() {
        let root = std::env::temp_dir().join("shard-free-name-test");
        let _ = std::fs::remove_dir_all(&root);
        std::fs::create_dir_all(&root).unwrap();
        let name = std::ffi::OsString::from("노래.webm");

        // Nothing there: the name it already has.
        assert_eq!(free_name(&root, &name), root.join("노래.webm"));

        // Something there: numbered, and the extension stays an extension.
        std::fs::write(root.join("노래.webm"), b"x").unwrap();
        assert_eq!(free_name(&root, &name), root.join("노래 (2).webm"));
        std::fs::write(root.join("노래 (2).webm"), b"x").unwrap();
        assert_eq!(free_name(&root, &name), root.join("노래 (3).webm"));

        let _ = std::fs::remove_dir_all(&root);
    }

    #[test]
    fn sizes_read_the_way_a_person_says_them() {
        assert_eq!(human(1 << 20), "1 MB");
        assert_eq!(human(3 << 30), "3.0 GB");
        assert_eq!(human(2048), "2 KB");
    }

    #[test]
    fn a_shelf_lists_what_is_in_its_folders_newest_first() {
        let dir = std::env::temp_dir().join(format!("shard-lib-test-{}", std::process::id()));
        let inner = dir.join("여행");
        std::fs::create_dir_all(&inner).expect("make test folders");
        std::fs::write(dir.join("첫번째.mp4"), b"aaaa").expect("write");
        std::fs::write(inner.join("두번째.mp4"), b"bb").expect("write");

        let mut found = Vec::new();
        sweep(&dir, "", Kind::Video, &mut found);
        assert_eq!(found.len(), 2);

        let top = found.iter().find(|i| i.title == "첫번째").expect("top item");
        assert_eq!(top.folder, "");
        assert_eq!(top.bytes, 4);
        let inside = found.iter().find(|i| i.title == "두번째").expect("foldered item");
        assert_eq!(inside.folder, "여행");

        let _ = std::fs::remove_dir_all(&dir);
    }
}
